metaflow/plugins/metadata/service.py

```diff
@@ -73,7 +94,55 @@ def register_task_id(self,
                            sys_tags=sys_tags)
         finally:
             self._register_code_package_metadata(run_id, step_name, task_id)
-        
+
+    def _start_heartbeat(self, heartbeat_type, flow_id, run_id, step_name=None, task_id=None):
+        if self._already_started():
+            raise Exception("heartbeat already started")
+        # start sidecar
+        if self.version() is None or \
+                LooseVersion(self.version()) < LooseVersion('2.0.4'):
+            # if local mode or old version of the service is running
+            # then avoid running real heartbeat sidecar process
+            self.sidecar_process = SidecarSubProcess("nullSidecarHeartbeat")
+        else:
+            self.sidecar_process = SidecarSubProcess("heartbeat")
+        # create init message
+        payload = {}
+        if heartbeat_type == HeartbeatTypes.TASK:
+            # create task heartbeat
+            data = {
+                    'flow_id': flow_id, 'run_number': run_id,
+                    "step_name": step_name, 'task_id': task_id,
+                    }
+            payload[self.HB_URL_KEY] = self.mli_url_task_template.format(**data)
+        elif heartbeat_type == HeartbeatTypes.RUN:
+            # create run heartbeat
+            data = {'flow_id': flow_id, 'run_number': run_id}
+
+            payload[self.HB_URL_KEY] = self.mli_url_run_template.format(**data)
+        else:
+            raise Exception("invalid heartbeat type")
+        payload["service_version"] = self.version()
+        msg = Message(MessageTypes.LOG_EVENT, payload)
+        self.sidecar_process.msg_handler(msg)
+
+    def start_run_heartbeat(self, flow_id, run_id):
+        self._start_heartbeat(HeartbeatTypes.RUN, flow_id, run_id)
+
+    def start_task_heartbeat(self, flow_id, run_id, step_name, task_id):
+        self._start_heartbeat(HeartbeatTypes.TASK,
+                              flow_id,
+                              run_id,
+                              step_name,
+                              task_id)
+
+    def _already_started(self):
+        return self.sidecar_process is not None
+
+    def stop_heartbeat(self):
+        msg = Message(MessageTypes.SHUTDOWN, None)
+        self.sidecar_process.msg_handler(msg)
```

Comment:
Should never be restarted but should we set sidecar_process to None at this time?

Reply:
I don't think setting it to None would be a good idea. Calling shutdown will send a message that will terminate the sidecar subprocess which will prevent it from sending more heartbeats. If we set sidecar_process to know then it looks like the sidecar never started and may cause confusion

Reply:
so once it is shutdown, if by mistake another message is sent to it, it will just be ignored?

Reply:
the message will just get blackholed since that PIPE is now broken, essentially the message will have no effect